File: kotlin-react-router/karakum/plugins/convertSkippedGenerics.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isQualifiedName(node.typeName)
        && ts.isIdentifier(node.typeName.right)
        && (
            node.typeName.right.text === "ReactElement"
            || node.typeName.right.text === "ComponentType"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }
    return null
}

File: kotlin-tanstack-virtual-core/karakum/src/jsMain/kotlin/wrappersgenerator/tanstack/virtual/core/plugins/convertSkippedGenerics.kt
package wrappersgenerator.tanstack.virtual.core.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.SyntaxKind
import typescript.asArray
import typescript.isIdentifier
import typescript.isTypeReferenceNode

// `Virtualizer<T, any>` / `Virtualizer<any, T>`
val convertSkippedGenerics = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Virtualizer")

        val typeArguments = ensureNotNull(node.typeArguments?.asArray())
        ensure(typeArguments.size == 2)
        ensure(typeArguments.any { it.kind == SyntaxKind.AnyKeyword })

        val arguments = typeArguments.map { argument ->
            if (argument.kind == SyntaxKind.AnyKeyword) "*" else render(argument)
        }

        "${render(typeName)}<${arguments.joinToString(", ")}>"
    }
}
